### PR TITLE
[1893] Implement FdSD private allowing for exchange for RAG

### PR DESCRIPTION
**Pull request by perwestling**
Exchange can be done in the following ways:
- To par RAG
- To exchange for up to 20% from Market

FdSD is closed when exchanging, or during the SR after first
5 train is bought in the OR before (directly after player has
had a chance to do the exchange).

### CODE REVIEW

File: lib/engine/game/g_1893/step/buy_sell_par_shares.rb [hunk 1]
```diff
@@ -2,23 +2,26 @@
...
           include BuyMinor
 
           FIRST_SR_ACTIONS = %w[buy_company pass].freeze
+          EXCHANGE_ACTIONS = %w[buy_shares].freeze
 
           def actions(entity)
+            return EXCHANGE_ACTIONS if entity == @game.fdsd && @game.rag.ipoed
             return [] unless entity&.player?
 
+            return %w[assign pass] if available_subsidiaries(entity).any?
```
Comment (tobymao) [inline lib/engine/game/g_1893/step/buy_sell_par_shares.rb:20]: unless empty?

File: lib/engine/game/g_1893/step/buy_sell_par_shares.rb [hunk 2]
```diff
@@ -97,6 +99,183 @@ def process_sell_shares(action)
 
             track_action(action, action.bundle.corporation)
           end
+
+          def process_par(action)
+            corporation = action.corporation
+            @par_rag = (action if rag_exchangable(action.entity, corporation))
+
+            return par_corporation(action) unless available_subsidiaries(current_entity).any?
```
Comment (tobymao) [inline lib/engine/game/g_1893/step/buy_sell_par_shares.rb:107]: if empty?

File: lib/engine/game/g_1893/step/buy_sell_par_shares.rb [hunk 3]
```diff
@@ -97,6 +99,183 @@ def process_sell_shares(action)
...
+                                        exchange_price: 0,
+                                        allow_president_change: true)
+
+            track_action(action, corporation)
+            @round.last_to_act = player
+            @round.current_actions << action
+          end
+
+          def description
+            return 'Exchange FdSD' if available_subsidiaries(current_entity).any?
```
Comment (tobymao) [inline lib/engine/game/g_1893/step/buy_sell_par_shares.rb:227]: unless empty

File: lib/engine/game/g_1893/step/buy_sell_par_shares.rb [hunk 4]
```diff
@@ -97,6 +99,183 @@ def process_sell_shares(action)
...
+
+            super
+          end
+
+          def pass_description
+            return 'Pass (Exchange FsDF)' if available_subsidiaries(current_entity).any?
```
Comment (tobymao) [inline lib/engine/game/g_1893/step/buy_sell_par_shares.rb:233]: unless empty

File: lib/engine/game/g_1893/step/buy_sell_par_shares.rb [hunk 5]
```diff
@@ -97,6 +99,183 @@ def process_sell_shares(action)
...
+
+            super
+          end
+
+          def log_pass(entity)
+            return if available_subsidiaries(current_entity).any?
```
Comment (tobymao) [inline lib/engine/game/g_1893/step/buy_sell_par_shares.rb:239]: unless empty

**Comment by perwestling** [comment]
@tobymao Have now fixed your 5 review comments. Have started at least one pre-alpha testing game (which includes this commit). I expect to put 1893 in alpha within a week or two. Was going to let Marflow Games check it out first as well.

**Pull request merged**